File: 08-craft/lab/api/app/shallow/repository.py
"""Layer 3 of 4. Declares `OrderRecord`. Forwards to the dao.

This is the layer Ousterhout's "pass-through method" describes exactly: the
signature is nearly identical to the one it calls, so there are now two places
to look, two names for one idea, and two things to keep in sync.

The stated justification -- "decouple from the ORM" -- does not survive reading
the file: SQLAlchemy's `Session` already IS that abstraction, and this layer
re-exposes it verbatim.
"""
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime

from sqlalchemy.ext.asyncio import AsyncSession

from .dao import OrderRow, select_order_count, select_orders


@dataclass(frozen=True)
class OrderRecord:
    """DTO #2 of 3. Structurally identical to OrderRow. Declared anyway."""

    id: int
    customer_id: int | None
    status: str
    total_cents: int
    created_at: datetime

    @classmethod
    def from_row(cls, row: OrderRow) -> "OrderRecord":
        return cls(
            id=row.id, customer_id=row.customer_id, status=row.status,
            total_cents=row.total_cents, created_at=row.created_at,
        )


async def fetch_orders(
    session: AsyncSession, customer_id: int, limit: int, offset: int
) -> list[OrderRecord]:
    rows = await select_orders(session, customer_id, limit, offset)
    return [OrderRecord.from_row(r) for r in rows]


async def count_orders(session: AsyncSession, customer_id: int) -> int:
    return await select_order_count(session, customer_id)

File: 08-craft/lab/probes/rust/shallow/src/repository.rs
//! Layer 2 of 3. Declares `OrderRecord`. Forwards to the dao.
//!
//! Read the two signatures side by side. `fetch_orders` repeats `&Store`,
//! `u64`, `usize` and `Result<_, OrderError>` exactly -- the compiler made the
//! pass-through impossible to disguise, which is the property topic 1 is
//! borrowing this language for.

use crate::dao::{select_order_count, select_orders, OrderRow};
use crate::{OrderError, Store};

/// DTO #2 of 3. Structurally identical to `OrderRow`. Declared anyway.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct OrderRecord {
    pub id: u64,
    pub customer_id: u64,
    pub status: String,
    pub total_cents: i64,
    pub created_at: i64,
}

impl From<OrderRow> for OrderRecord {
    fn from(r: OrderRow) -> Self {
        OrderRecord {
            id: r.id,
            customer_id: r.customer_id,
            status: r.status,
            total_cents: r.total_cents,
            created_at: r.created_at,
        }
    }
}

pub fn fetch_orders(
    store: &Store,
    customer_id: u64,
    limit: usize,
) -> Result<Vec<OrderRecord>, OrderError> {
    Ok(select_orders(store, customer_id, limit)?
        .into_iter()
        .map(OrderRecord::from)
        .collect())
}

pub fn count_orders(store: &Store, customer_id: u64) -> usize {
    select_order_count(store, customer_id)
}
